### PR TITLE
add new grid component

### PR DESCRIPTION
**Pull request by sstraatemans**
- [X] I have reviewed my changes and run the appropriate tests.
- [X] I have have run `rush change` to add the appropriate change logs.
- [ ] I have added/edited docs.
- [ ] I have added tutorials.
- [ ] I have double checked and DEFINITELY added docs.

### CODE REVIEW

File: packages/libs/react-components/src/components/Grid/GridItem.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  children?: ReactNode;
+  colStart?: IColSpan;
+  colEnd?: IColSpan;
+  rowStart?: IColSpan;
+  rowEnd?: IColSpan;
+  area?: string;
+  bg?: IThemeColors;
+}
+
+const checkSpan = (
```
Comment (eileenmguo) [inline packages/libs/react-components/src/components/Grid/GridItem.tsx:21]: What is this checking? Should it be validating that the colEnd is past colStart?

File: packages/libs/react-components/src/styles/stitches.config.ts [hunk 2]
```diff
@@ -23,7 +24,16 @@ const sizes: Record<string, string> = {
...
+} as const;
+
+// eslint-disable-next-line @kadena-dev/typedef-var
+export const breakpoints = {
+  sm: `(min-width: ${640 / 16}rem)`,
+  md: `(min-width: ${768 / 16}rem)`,
+  lg: `(min-width: ${1024 / 16}rem)`,
+  xl: `(min-width: ${1280 / 16}rem)`,
+  '2xl': `(min-width: ${1536 / 16}rem)`,
+} as const;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/styles/stitches.config.ts:36]: What is the reasoning for exporting these?

File: packages/libs/react-components/src/typings/index.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { IThemeBreakpointProps } from '../styles';
+
+export type Primitive = number | string | boolean;
+export type ICSSPropType =
+  | Primitive
+  | Partial<Record<IThemeBreakpointProps, Primitive>>;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/typings/index.ts:6]: Is there a reason you opted for `color: { "sm": "$blue500", "md": "$gray500"}` instead of `color: { "@sm": "$blue500", "@md": "$gray500"}`?

The reason I ask is because I think the `@` symbol makes it very clear that we are referring to a breakpoint and this is a style `at size sm`. It also would help us reduce the number of types

Comment (eileenmguo) [reply packages/libs/react-components/src/typings/index.ts:6]: Additionally, I think it would be valuable to make this type more strict by making it a generic so that we can pass in the property we will apply it to. 
For example: backgroundColor 
`ICSSPropType<Stitches.PropertyValue<"backgroundColor">>`

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+const GridContainer: FC<IGridContainerProps> = ({
+  children,
+  gap,
+  templateAreas,
+  templateRows,
+  templateColumns,
+}) => {
+  const styles = convertStyles({
+    gridGap: gap ?? {},
```
Comment (eileenmguo) [inline packages/libs/react-components/src/components/Grid/GridContainer.tsx:30]: Is this necessary? My understanding is that gap can only be undefined or the value itself so would this work if it were just written as the following? `gridGap: gap,`

File: packages/libs/react-components/src/utils/convertStyles.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { Primitive } from '../typings';
+
+export const convertStyles = (
+  styles: Record<string, Primitive | Record<string, Primitive>>,
+): Record<string, Primitive | Record<string, Primitive>> => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const newStyles = {} as any;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/utils/convertStyles.ts:7]: can we use unknown here?

File: packages/libs/react-components/src/utils/convertStyles.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    if (value === undefined) return;
+
+    if (typeof value === 'object') {
+      return Object.keys(value).forEach((mediaQueryKey) => {
+        if (newStyles['@' + mediaQueryKey] === undefined) {
+          newStyles['@' + mediaQueryKey] = {};
+        }
+
+        const prop = styles[key];
+        if (typeof prop === 'object') {
```
Comment (eileenmguo) [inline packages/libs/react-components/src/utils/convertStyles.ts:18]: I'm a bit confused. Isn't `prop` the same as `value` so it would be of type "object" by default?

File: packages/libs/react-components/src/utils/convertStyles.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+          newStyles['@' + mediaQueryKey][key] = prop[mediaQueryKey];
+        }
+      });
+    }
+    return (newStyles[key] = value);
+  });
+
+  return newStyles;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/utils/convertStyles.ts:26]: I noticed this doesn't account for sizes smaller than what is defined. I think it would be beneficial to implement an `@initial` like stitches does so they can also set a default value
https://stitches.dev/docs/responsive-styles#setting-an-initial-variant

File: packages/libs/react-components/tsconfig.json [hunk 8]
```diff
@@ -2,20 +2,7 @@
...
-    "outDir": "dist/esm",
-    "allowJs": true,
-    "moduleResolution": "node",
-    "allowSyntheticDefaultImports": true,
-    "forceConsistentCasingInFileNames": true
-  },
-  "include": ["src/**/*"]
+    "outDir": "lib"
+  }
 }
```
Comment (eileenmguo) [inline packages/libs/react-components/tsconfig.json:8]: what was the reasoning for removing this? I believe this change will break the usage of this package since package.json is still pointing at `dist/esm`

**Review by eileenmguo — CHANGES_REQUESTED** [review]
I noticed this PR doesn't take the design system spacing into account. Is there a reason for that? I think it would be beneficial to offer a variant that allows users to choose a predefined space from the design system so they don't have to think about which of the theme spaces aligns with the design. 

This is something that would be shared with the stack component so I can also just implement that in my PR and apply it to the grid component as well. Just wanted to check if there was a rational for not including it!

File: packages/libs/react-components/src/utils/convertStyles.ts (hunk 5 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/utils/convertStyles.ts:7]: it will break the function.
Albert and I took the best part of the morning to get this function properly typed. we failed. :/

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx (hunk 4 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridContainer.tsx:30]: hmm..thinking about it again. I added undefined to the Primitive type (used in the convertStyles function)
makes it much better. thanks

File: packages/libs/react-components/src/styles/types.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { colors } from './colors';
+import { breakpoints, sizes } from './stitches.config';
+
+export type IThemeColors = `$${keyof typeof colors}`;
+export type IThemeSizes = `$${keyof typeof sizes}`;
+export type IThemeBreakpointProps = keyof typeof breakpoints;
+export type IStyleBreakpointProps = `@${keyof typeof breakpoints}`;
```
[the remark this answers is not in the document]
Comment (sstraatemans) [reply packages/libs/react-components/src/styles/types.ts:7]: makes sense

File: packages/libs/react-components/src/styles/stitches.config.ts (hunk 2 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/styles/stitches.config.ts:36]: for the typing. 
These are not in the theme object. so need them exported

File: packages/libs/react-components/src/components/Grid/GridItem.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+export type IColSpan = number | 'auto';
+
+export interface IGridItemProps {
+  children?: ReactNode;
+  colStart?: IColSpan;
+  colEnd?: IColSpan;
+  rowStart?: IColSpan;
+  rowEnd?: IColSpan;
+  area?: string;
+  bg?: IThemeColors;
```
[the remark this answers is not in the document]
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:18]: but then we use all the colors. that is really flexible

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 1 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:21]: we can add that check. 
But is is really just creating that string.

File: packages/libs/react-components/src/utils/convertStyles.ts (hunk 6 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/utils/convertStyles.ts:18]: it didnt' but now that you mentioned it, i refactored and it looks nicer now

File: packages/libs/react-components/src/utils/convertStyles.ts (hunk 7 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/utils/convertStyles.ts:26]: shouldnt the initial be in the component it self already?

File: packages/libs/react-components/tsconfig.json (hunk 8 above)
Comment (sstraatemans) [reply packages/libs/react-components/tsconfig.json:8]: I have no idea...I was desperately trying to get the build to run

File: packages/libs/react-components/src/utils/convertStyles.ts (hunk 5 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/utils/convertStyles.ts]: not needed anymore

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx (hunk 4 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridContainer.tsx]: not needed anymore

File: packages/libs/react-components/src/components/Grid/GridItem.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import { IThemeColors, styled } from './../../styles';
+
+import React, { FC, ReactNode } from 'react';
+
+const BaseItem = styled('div', {
+  gridColumnStart: 'auto',
+});
+
+export type IColSpan = number | 'auto';
```
[the remark this answers is not in the document]
Comment (ash-vd) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:10]: I think we should only prepend an I when it's an actual interface. If one wants to extend it because "it's an interface" it doesn't work. We could do `ColSpanType` if you're not happy with just `ColSpan` (but I think `ColSpan` is fine

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ash-vd) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:19]: This is a tricky one. Is the general idea that we only allow the user to use colors from the theme? In my opinion it wouldn't hurt to be flexible here and allow all colors to be used.

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  spacing?: Stitches.VariantProps<typeof BasicContainer>['spacing'];
+  templateRows?: Stitches.CSSProperties['gridTemplateRows'];
+  templateColumns?: Stitches.CSSProperties['gridTemplateColumns'];
+  templateAreas?: Stitches.CSSProperties['gridTemplateAreas'];
+}
+
+const BasicContainer = styled('div', {
+  display: 'grid',
+  gridTemplateColumns: 'repeat(12, 1fr)',
+  gridGap: '$md',
```
Comment (eileenmguo) [inline packages/libs/react-components/src/components/Grid/GridContainer.tsx:17]: This is a matter of preference, but I think it's a bit clearer to define the default gridGap using the default variant

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:10]: Yeah, I also think it's better to only prefix it if it is an Interface as it indicates that it is extendable. My preference is to go with `ColSpan`, but `ColSpanType` is also fine

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:19]: Yeah, this one I'm a bit torn on, but I'm leaning towards providing the flexibility. I think it's clear that we should only be using colors from the theme, but I also feel that we can trust our team to use this correctly. I'm fine with either approach at this point

File: packages/libs/react-components/src/styles/index.ts [hunk 13]
```diff
@@ -7,6 +7,8 @@ export {
   getCssText,
   createTheme,
   globalCss,
+  sizes,
```
Comment (eileenmguo) [inline packages/libs/react-components/src/styles/index.ts:10]: Why do we need to export sizes?

File: packages/libs/react-components/tsconfig.json [hunk 14]
```diff
@@ -2,20 +2,7 @@
...
-    "jsx": "react",
-    "module": "ESNext",
-    "declaration": true,
-    "declarationDir": "types",
-    "sourceMap": true,
-    "outDir": "dist/esm",
-    "allowJs": true,
-    "moduleResolution": "node",
-    "allowSyntheticDefaultImports": true,
-    "forceConsistentCasingInFileNames": true
```
Comment (eileenmguo) [inline packages/libs/react-components/tsconfig.json:18]: I would revert this change since this is this configuration is what transpiles the esm version of the lib

**Review by eileenmguo — CHANGES_REQUESTED** [review]

File: packages/libs/react-components/src/styles/index.ts (hunk 13 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/styles/index.ts:10]: removed

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import { styled } from './../../styles';
+
+import type * as Stitches from '@stitches/react';
```
[the remark this answers is not in the document]
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridContainer.tsx:3]: done

File: packages/libs/react-components/src/components/Grid/Grid.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+import type { Meta, StoryObj } from '@storybook/react';
+import React from 'react';
+
+const meta: Meta<typeof Grid.Container> = {
+  title: 'Grid',
+  component: Grid.Container,
+  argTypes: {
+    spacing: {
+      options: Object.keys(theme.sizes).map((key) => key),
```
[the remark this answers is not in the document]
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/Grid.stories.tsx:13]: Ohh actually I think this should be taken from the variants since we are restricting it only be able to use 2xs, xs, sm...?

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:10]: I like the ColsSpanType

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:19]: fixed this

File: packages/libs/react-components/src/components/Grid/Grid.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/Grid.stories.tsx:13]: > Ohh actually I think this should be taken from the variants since we are restricting it only be able to use 2xs, xs, sm...?

👀

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx (hunk 12 above)
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/GridContainer.tsx:17]: You should be able to remove `gridGap: "$md"` now that the default variant is defined

File: packages/libs/react-components/src/components/Grid/GridItem.tsx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+export type ColSpanType = number | 'auto';
+
+export interface IGridItemProps {
+  children?: ReactNode;
+  colStart?: ColSpanType;
+  colEnd?: ColSpanType;
+  rowStart?: ColSpanType;
+  rowEnd?: ColSpanType;
+  area?: CSSProperties['gridArea'];
+  bg: PropertyValue<'backgroundColor', typeof config>;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/components/Grid/GridItem.tsx:21]: 🥳

File: packages/libs/react-components/src/components/Grid/GridItem.tsx [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+  rowStart,
+  rowEnd,
+  area,
+  bg,
+}) => {
+  const styles = {
+    gridColumn: createSpan(colStart, colEnd),
+    gridRow: createSpan(rowStart, rowEnd),
+    gridArea: area,
+    backgroundColor: bg,
```
Comment (eileenmguo) [inline packages/libs/react-components/src/components/Grid/GridItem.tsx:45]: maybe it would be good to add this util to stitches so we can always use `bg` to define backgroundColor? What do you think?

File: packages/libs/react-components/src/styles/colors/index.ts [hunk 19]
```diff
@@ -1,13 +1,14 @@
 // TODO: Update to match design system
-export const colors: Record<string, string> = {
+// eslint-disable-next-line @kadena-dev/typedef-var
+export const colors = {
```
Comment (eileenmguo) [inline packages/libs/react-components/src/styles/colors/index.ts:3]: Should we add the type back so we can remove the eslint disable?

File: packages/libs/react-components/src/styles/stitches.config.ts [hunk 20]
```diff
@@ -4,8 +4,10 @@ import { colors, colorsDark } from './colors';
 
 import type { CSS as StitchesCSS } from '@stitches/react';
 import { createStitches, PropertyValue } from '@stitches/react';
+import { ConfigType } from '@stitches/react/types/config';
 
-const sizes: Record<string, string> = {
+// eslint-disable-next-line @kadena-dev/typedef-var
```
Comment (eileenmguo) [inline packages/libs/react-components/src/styles/stitches.config.ts:9]: Should we add the type back here too? I think it would be better to use eslint-disable as little as possible

File: packages/libs/react-components/src/styles/stitches.config.ts [hunk 21]
```diff
@@ -169,4 +173,6 @@ export const darkTheme = createTheme('darkTheme', {
   },
 });
 
+export type IThemeColors = `$${keyof typeof theme.colors}`;
```
Comment (eileenmguo) [inline packages/libs/react-components/src/styles/stitches.config.ts:176]: Would it be worth using ScaleValue here instead of composing it ourselves?

File: packages/libs/react-components/tsconfig.test.json [hunk 22]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (eileenmguo) [inline packages/libs/react-components/tsconfig.test.json:1]: nice! glad this works! 

Some things that still need to change: 
- I think it would be better to name this tsconfig.esm.json since this transpiles the lib for esm
- We need to update the package.json `build` script from 
```"build": "tsc & tsc -p ./tsconfig.cjs.json",```
to
```"build": "tsc -p ./tsconfig.esm.json & tsc -p ./tsconfig.cjs.json",```

Comment (eileenmguo) [reply packages/libs/react-components/tsconfig.test.json:1]: Also the `tsconfig.cjs.json` will need to extend the `tsconfig.esm.json` since it was previously extending that tsconfig and overwriting certain values

File: packages/libs/react-components/src/styles/stitches.config.ts (hunk 21 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/styles/stitches.config.ts:176]: for now we dont need this import anymore.
will remove it

File: packages/libs/react-components/src/styles/stitches.config.ts (hunk 20 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/styles/stitches.config.ts:9]: yes

File: packages/libs/react-components/tsconfig.test.json (hunk 22 above)
Comment (sstraatemans) [reply packages/libs/react-components/tsconfig.test.json:1]: done

File: packages/libs/react-components/src/components/Grid/GridItem.tsx (hunk 18 above)
Comment (sstraatemans) [reply packages/libs/react-components/src/components/Grid/GridItem.tsx:45]: We need to config inside the config, which is not possible it seems.
lets look at this later

**Review by isaozler — APPROVED** [review]

File: packages/libs/react-components/src/styles/colors/index.ts (hunk 19 above)
Comment (eileenmguo) [reply packages/libs/react-components/src/styles/colors/index.ts:3]: can we remove the eslint-disable?

File: packages/libs/react-components/src/components/Grid/GridContainer.tsx (hunk 12 above)
Comment (eileenmguo) [reply packages/libs/react-components/src/components/Grid/GridContainer.tsx:17]: not sure if you saw this since the original comment was resolved

**Review by eileenmguo — APPROVED** [review]
🎉
